src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -107,13 +107,89 @@ bool RaftReplDev::join_group() {
         m_msg_mgr.join_group(m_group_id, "homestore_replication",
                              std::dynamic_pointer_cast< nuraft_mesg::mesg_state_mgr >(shared_from_this()));
     if (!raft_result) {
-        HS_DBG_ASSERT(false, "Unable to join the group_id={} with error={}", boost::uuids::to_string(m_group_id),
-                      raft_result.error());
+        HS_DBG_ASSERT(false, "Unable to join the group_id={} with error={}", group_id_str(), raft_result.error());
         return false;
     }
     return true;
 }
 
+AsyncReplResult<> RaftReplDev::replace_member(replica_id_t member_out_uuid, replica_id_t member_in_uuid) {
+    LOGINFO("Replace member group_id={} member_out={} member_in={}", group_id_str(),
+            boost::uuids::to_string(member_out_uuid), boost::uuids::to_string(member_in_uuid));
+
+    // Check if member_out exists in the group.
+    std::vector< nuraft::ptr< nuraft::srv_config > > configs_out;
+    m_msg_mgr.get_srv_config_all(m_group_id, configs_out);
+    if (std::find_if(configs_out.begin(), configs_out.end(), [member_out_uuid](const auto& cfg) {
+            return boost::uuids::to_string(member_out_uuid) == cfg->get_endpoint();
+        }) == configs_out.end()) {
+        RD_LOGE("Member out not found in the group.");
+        return make_async_error<>(ReplServiceError::SERVER_NOT_FOUND);
+    }
+
+    // Step 1: Check if leader itself is requested to move out.
+    if (m_my_repl_id == member_out_uuid && m_my_repl_id == get_leader_id()) {
```

Comment:
One issue I see is: when SM experienced disk down, if it happens on leader, HomeStore should yield leadership immediately, the reason is, if we rely on CM to do that, there needs to be some manual ops to kicks in or even if it is automated from CM layer, it also takes some time for CM to detect that. If Homestore can yield leadership it to other member, the other two copies can serve traffic immeidately without any SOPs need to be taken in production.

Reply:
Created an issue for this https://github.com/eBay/HomeStore/issues/563